Have a from-zero-to-hero pack to build PyTorch from scratch

lezcano opened the pull request:
This PR adds a README, a few short utility files, and merges
`torch-common` and `torch-common-nebari` in favour of a fully-commented
file.

The diffs are terrible, better to look at the actual files to review.

lezcano committed:
Have a from-zero-to-hero pack to build PyTorch from scratch

This PR adds a README, a few short utility files, and merges
`torch-common` and `torch-common-nebari` in favour of a fully-commented
file.

I tested this workflow in paperspace and it all seems to work end-to-end.

lezcano committed:
git rm and simplify a couple things

lezcano commented on pytorch-dev.yaml:
```diff
@@ -25,16 +24,19 @@ dependencies:
+  - pytorch::magma-cuda118
+  - conda-forge::cuda-version=11.8
+  - conda-forge::cudnn
+  - cuda-libraries-dev
   - cuda-nvcc
   - cuda-gdb
+  - cuda-nvtx
+  - cuda-nvml-dev
+  - cuda-cupti
```
According to [NVIDIA's docs](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#installing-cuda-using-conda), it seems like CUDA toolkit is not recommended. They suggest using individual packages. No idea why, as they still distribute `cuda-toolkit` in conda.

amjames replied:
I also have git in the env recipe as the default git version on some systems is too old and ghstack will not work with it.

peterbell10 commented on pytorch-build.sh:
```diff
@@ -5,7 +5,11 @@ eval "$(conda shell.bash hook)"
 
-source ~/git/torch-common.sh
+
+source ./torch-common.sh
```
Does this work even though you're cd'ed into the pytorch directory?

lezcano replied:
This is wrong. I changed it because I thought it'd work but it doesn't. Let me roll back.

peterbell10 commented on torch-clone.sh:
```diff
@@ -0,0 +1,8 @@
+mkdir -p ~/git
+cd ~/git
+for pkg in ${PKGS[@]}; do
+	git clone git@github.com:pytorch/${pkg}.git
```
Why not use https? Also won't this clone into `audio` not `torch-audio`.

lezcano replied:
but it's audio, isn't it? https://github.com/pytorch/audio

peterbell10 replied:
Yes, but 
1. `audio` without qualification isn't a very descriptive name
2. the `torch-build.sh` script refers to the `torch-audio` directory

peterbell10 commented on torch-common.sh:
```diff
@@ -1,53 +1,42 @@
-set -e
 
-eval "$(conda shell.bash hook)"
-conda activate pytorch-dev
```
The idea of having `conda activate is that you can't inadvertently build pytorch in the wrong environment.

lezcano replied:
The thing is that this is just flags, it shouldn't even be executed on its own. We have this in `torch-build.py` and `pytorch-build.py`

peterbell10 replied:
`torch-build.sh` doesn't activate because it relies on `torch-common.sh` for that. I agree the duplication in `pytorch-build.sh` is unnecessary though, that one can be removed.

peterbell10 commented on torch-common.sh:
```diff
@@ -1,53 +1,42 @@
+export MAX_JOBS=${MAX_JOBS:-95}
+export CMAKE_BUILD_TYPE=Release
+# CMAKE_BUILD_TYPE=RelWithDebInfo gives you line numbers on gdb,
+# but makes the symbol loading phase in gdb and the linking phase in compilation much slower.
```
Are you suggesting that not having symbols makes gdb earier to use???

lezcano replied:
I'm exposing the pros and cons. Then everyone should use the one that works for them I guess.

peterbell10 replied:
Yes but the default matters because otherwise when you need symbols you have to rebuild pytorch from scratch. I can see the argument for link-time improvements, but saying you get a better experience in gdb without debug info seems iffy to me.

lezcano replied:
I don't think it says the debug experience is better, it just says that the symbols take longer to load. It also says it gives you line numbers. Each can choose.

I think that, a priori, the default should be this one, as if someone does not use gdb, they would be getting the fast compilation by default. Then, if they want to use gdb, they can change it if they want to use line numbers.

The way I see this is that we could have a repo, and then people could have their own branch with the changes they prefer. In this setting, as long as everything is commented and the defaults are not bonkers it should be alright.

peterbell10 commented on torch-common.sh:
```diff
@@ -1,53 +1,42 @@
+export BUILD_TEST=0                               # C++ tests
+export BUILD_CAFFE2=0                             # caffe2
+export BUILD_CAFFE2_OPS=0                         # caffe2
+export USE_DISTRIBUTED=0                          # distributed
```
I get errors running inductor if I don't build with distributed.

lezcano replied:
It may need a full rebuild. I'm building with these flags in qgpu3 and it works just fine.

peterbell10 replied:
Oh actually I just tried it and you're right. Maybe it was fixed since I last tried it.

peterbell10 commented on torch-update.sh:
```diff
@@ -0,0 +1,20 @@
+  git rebase origin/main
+  git submodule update --init --recursive --jobs 0
+  git prune
+  cd ..
```
`pushd` and `popd`?

lezcano replied:
yup, that's cleaner

peterbell10 commented on torch-common.sh:
```diff
@@ -1,53 +1,42 @@
 #!/bin/bash
-set -e
```
`set -e` maybe doesn't matter here, but I think it's a good default in all bash scripts. As if one command fails then it may break the scripts assumptions, so you usually don't want to continue running.

lezcano replied:
Same as above, this is just a file with some flags, not meant to be executed.

peterbell10 commented on torch-build.sh:
```diff
@@ -1,14 +1,16 @@
 #!/bin/bash
 
-cd ~/git
+bash ./pytorch-build.sh
```
Doesn't this create a new environment, so if you specify `MAX_JOBS=10` then it will be ignored?

lezcano replied:
they both source `torch-common.py`, so that should be alright. I didn't know whether doing `export` rather than `source` in `pytorch-build.sh` would allow for removing this repetition.

peterbell10 replied:
You can currently specify `MAX_JOBS` on the command line like:
```bash
MAX_JOBS=10 ~/git/torch-build.sh
```

But I think this change breaks it.

lezcano replied:
It does work and MAX_JOBS=10 in the inner script.

lezcano committed:
Address review

lezcano commented:
Addressed the reviews. Now we don't have repetition of anything, and since we source `pytorch-build.py` from `torch-build.py`, we inherit all the variables and the conda environment. Left a comment.

About the `Release` vs with debug symbols, if you write a different explanation, I'm happy to change it, but I still think that, by default, it may be better to leave it as `Release`, as most people don't use gdb in their day-to-day workflow.

lezcano commented:
cc @amjames this one should be ready to review

lezcano committed:
Update cores computation

lezcano committed:
Add torchbench instructions

lezcano committed:
Several fixes, add link c++, change torch-build to loop

lezcano committed:
minor

peterbell10 commented on torch-clone.sh:
```diff
@@ -0,0 +1,11 @@
+cd ~/git
+git clone git@github.com:pytorch/pytorch.git
+for pkg in ${PKGS[@]}; do
+	git clone git@github.com:pytorch/${pkg}.git "pytorch-${pkg}"
```
```suggestion
	git clone git@github.com:pytorch/${pkg}.git "torch-${pkg}"
```

peterbell10 commented on torch-clone.sh:
```diff
@@ -0,0 +1,11 @@
+
+mkdir -p ~/git
+cd ~/git
+git clone git@github.com:pytorch/pytorch.git
```
You never said why you prefer `git@` over `https://`. I guess it doesn't matter too much though?

lezcano replied:
I usually forward my ssh keys and `git@` is the way to use it. I though that `https://` was to log through password, which I believe it's disrecommended?

peterbell10 replied:
When you have 2FA enabled you need to use a personal access token instead of a password, but you need to generate tokens to use ghstack anyway so I don't really see it as an issue.

peterbell10 replied:
SSH key forwarding does sound like a pretty good reason to use `git@` though.

lezcano replied:
Let's have a chat on Monday, but I'm marking this one as resolved for now.

amjames commented on torch-common.sh:
```diff
@@ -1,53 +1,51 @@
+export USE_DISTRIBUTED=0                          # distributed
+export USE_NCCL=0                                 # distributed
+export USE_GLOO=0                                 # distributed
+export USE_XNNPACK=0                              # quantized
```
At some point I got the advice to include `USE_GOLD_LINKER=1`. Is that good to add here, or is that advice now out of date?

lezcano replied:
It's a bit of an optional point. I never got huge speed-ups from using it so I just stopped using it. WDYT @peterbell10?

lezcano replied:
I think that, in the interest of simplicity, we should leave it out for now. If it really is faster, we could potentially leave it commented out with a comment describing how to set it up.

peterbell10 replied:
IIRC I tried it before and might have had some compatibility issues with cupti. Don't quote me on that though.

amjames commented on pytorch-build.sh:
```diff
@@ -5,7 +5,12 @@ eval "$(conda shell.bash hook)"
 conda activate pytorch-dev
 cd ~/git/pytorch
```
these script do assume some higher level directory structure (which is completely reasonable and probably the way most people would have it set up), but it should probably be referenced in the README. 

Alternatively the paths could be edited to assume that rather than `~/git/<repo>` the repositories have the same parent as this one, (ie  `$SCRIPTS_DIR/../<repo>`), wherever a given user has decided to put their repos.

lezcano replied:
The level directory structure is given by the `torch-clone.sh` script. I'll leave a note in the README, that's a good idea.

amjames reviewed (approved):
This looks great and is going to help us out a lot! See inline for couple of suggestions, nothing blocking.

peterbell10 commented on torch-common.sh:
```diff
@@ -1,53 +1,51 @@
+export BUILD_TEST=0                               # C++ tests
+export BUILD_CAFFE2=0                             # caffe2
+export BUILD_CAFFE2_OPS=0                         # caffe2
+export USE_DISTRIBUTED=0                          # distributed
```
Actually I ran into the issue again and it's torchbench, not the tests that fail. I get this error:
```
pbell@qgpu3:~/git/pytorch$ conda activate pytorch-dev && PYTHONPATH=$HOME/git/pytorch-benchmark/ python benchmarks/dynamo/torchbench.py 
Traceback (most recent call last):
  File "benchmarks/dynamo/torchbench.py", line 16, in <module>
    from common import BenchmarkRunner, main
  File "/home/pbell/git/pytorch/benchmarks/dynamo/common.py", line 39, in <module>
    from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
  File "/home/pbell/git/pytorch/torch/distributed/fsdp/__init__.py", line 1, in <module>
    from .flat_param import FlatParameter
  File "/home/pbell/git/pytorch/torch/distributed/fsdp/flat_param.py", line 30, in <module>
    from torch.distributed._tensor import DTensor
  File "/home/pbell/git/pytorch/torch/distributed/_tensor/__init__.py", line 6, in <module>
    import torch.distributed._tensor.ops
  File "/home/pbell/git/pytorch/torch/distributed/_tensor/ops/__init__.py", line 2, in <module>
    from .embedding_ops import *  # noqa: F403
  File "/home/pbell/git/pytorch/torch/distributed/_tensor/ops/embedding_ops.py", line 6, in <module>
    from torch.distributed._tensor.api import _Partial, DTensorSpec, Replicate, Shard
  File "/home/pbell/git/pytorch/torch/distributed/_tensor/api.py", line 8, in <module>
    import torch.distributed._tensor.dispatch as op_dispatch
  File "/home/pbell/git/pytorch/torch/distributed/_tensor/dispatch.py", line 10, in <module>
    from torch.distributed._tensor.device_mesh import DeviceMesh
  File "/home/pbell/git/pytorch/torch/distributed/_tensor/device_mesh.py", line 6, in <module>
    import torch.distributed._functional_collectives as funcol
  File "/home/pbell/git/pytorch/torch/distributed/_functional_collectives.py", line 6, in <module>
    import torch.distributed.distributed_c10d as c10d
  File "/home/pbell/git/pytorch/torch/distributed/distributed_c10d.py", line 17, in <module>
    from torch._C._distributed_c10d import (
ModuleNotFoundError: No module named 'torch._C._distributed_c10d'; 'torch._C' is not a package

```

lezcano replied:
This looks like an issue tat should be fixed in master? All the distributed imports should be guarded...

lezcano replied:
Filled an issue https://github.com/pytorch/pytorch/issues/102086.

lezcano replied:
Fixed in https://github.com/pytorch/pytorch/pull/103163

amjames commented on README.md:
```diff
@@ -0,0 +1,56 @@
+# Compiling PyTorch
+
+If you don't have gcc, g++, install them from apt-get.
+If you want to go the extra mile, you can install gcc-9, g++-9, as these are the ones that are the versions used in CI (as of this writing). For example, I did
```
FYI https://github.com/pytorch/pytorch/issues/101839 will make gcc-9 the min supported version. From the triage meeting there does not seem to be any objection and that change will go through here at some point. So the gcc-9 steps would become a requirement rather than just a good idea.

lezcano replied:
Yeah, both the gcc and the python versions are set to the min required, to avoid using newer unsupported features from newer compilers.

I'll add a note on this in the readme.

lezcano committed:
pytorch- -> torch-. Add a few more comments to the README

peterbell10 commented on pytorch-build.sh:
```diff
@@ -5,7 +5,12 @@ eval "$(conda shell.bash hook)"
 
-source ~/git/torch-common.sh
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
```
I think this needs to be the first thing done in the script. Currently if I call the script as `./pytorch-build.sh` then `SCRIPT_DIR` is `~/git/pytorch` because `dirname ${BASH_SOURCE[0]}` is just `.` and we've changed the current directory above.

lezcano replied:
uh, good catch!

lezcano replied:
fixed

lezcano committed:
Fix SCRIPT_DIR

peterbell10 commented on torch-build.sh:
```diff
@@ -1,16 +1,24 @@
-pip uninstall -y torch functorch torchvision torchtext torchaudio torchdata
+pip uninstall -y torchvision torchtext torchaudio torchdata torchbenchmark
+
+PKGS=(data vision text audio benchmark)
```
Torchbench doesn't have a `setup.py`
```suggestion
PKGS=(data vision text audio)
```

lezcano committed:
Update torch-build.sh

Co-authored-by: peterbell10 <peterbell10@live.co.uk>
